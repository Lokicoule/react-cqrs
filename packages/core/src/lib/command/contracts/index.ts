import { Unsubscribe } from '../../types';

export interface CommandContract {
  readonly commandName: string;
}

export interface CommandHandlerContract<
  TCommand extends CommandContract = CommandContract,
  TResult = unknown
> {
  execute(command: TCommand): TResult;
}

export interface CommandBusContract {
  register<TCommand extends CommandContract = CommandContract>(
    command: TCommand,
    handler: CommandHandlerContract<TCommand>
  ): Unsubscribe;

  execute<TCommand extends CommandContract, TResult = unknown>(
    command: TCommand
  ): TResult;
}
