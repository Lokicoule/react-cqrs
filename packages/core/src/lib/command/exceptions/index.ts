export * from './CommandAlreadyRegisteredException';
export * from './CommandNotFoundException';
