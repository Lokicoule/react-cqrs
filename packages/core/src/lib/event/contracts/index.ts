import { Unsubscribe } from '../../types';
import { EventResult } from '../types';

/**
 * @name EventContract
 * @property eventName - The identifier of the event used to register the event handler.
 */
export interface EventContract {
  readonly eventName: string;
}

/**
 * @name EventHandlerContract
 * @method execute - Executes the event handler.
 */
export interface EventHandlerContract<
  TEvent extends EventContract = EventContract
> {
  handle(event: TEvent): EventResult;
}

/**
 * @name EventBusContract
 * @method register - Registers a event handler for a event.
 * @method handle - Handles the event.
 */
export interface EventBusContract {
  register<TEvent extends EventContract = EventContract>(
    event: TEvent,
    handler: EventHandlerContract<TEvent>
  ): Unsubscribe;

  handle<TEvent extends EventContract>(event: TEvent): EventResult;
}
