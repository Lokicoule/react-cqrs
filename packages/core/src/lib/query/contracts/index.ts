import { Unsubscribe } from '../../types';

export interface QueryContract {
  readonly queryName: string;
}

export interface QueryHandlerContract<
  TQuery extends QueryContract = QueryContract,
  TResult = unknown
> {
  execute(query: TQuery): TResult;
}

export interface QueryBusContract {
  register<TQuery extends QueryContract = QueryContract>(
    query: TQuery,
    handler: QueryHandlerContract<TQuery>
  ): Unsubscribe;

  execute<TQuery extends QueryContract, TResult = unknown>(
    query: TQuery
  ): TResult;
}
