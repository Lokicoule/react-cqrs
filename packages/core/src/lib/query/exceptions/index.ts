export * from './QueryAlreadyRegisteredException';
export * from './QueryNotFoundException';
