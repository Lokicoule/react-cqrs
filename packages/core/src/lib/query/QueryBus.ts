import { Unsubscribe } from '../types';
import { QueryBusContract, QueryContract } from './contracts';
import {
  QueryAlreadyRegisteredException,
  QueryNotFoundException,
} from './exceptions';
import { QueryHandler } from './types';

export class QueryBus implements QueryBusContract {
  private readonly handlers = new Map<string, QueryHandler>();

  public register<TQuery extends QueryContract>(
    query: TQuery,
    handler: QueryHandler<TQuery>
  ): Unsubscribe {
    if (this.handlers.has(query.queryName)) {
      throw new QueryAlreadyRegisteredException(query.queryName);
    }

    this.handlers.set(query.queryName, handler as QueryHandler);

    return () => this.handlers.delete(query.queryName);
  }

  public execute<TQuery extends QueryContract, TResult = void>(
    query: TQuery
  ): TResult {
    const handler = this.handlers.get(query.queryName);

    if (!handler) {
      throw new QueryNotFoundException(query.queryName);
    }

    if (typeof handler === 'function') {
      return handler(query) as TResult;
    }

    return handler.execute(query) as TResult;
  }
}
