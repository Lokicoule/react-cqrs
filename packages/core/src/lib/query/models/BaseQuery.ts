import { QueryContract } from '../contracts';

/**
 * @name BaseQuery
 * @description This base class helps to enforce the query name to be a static property.
 * @property queryName - The identifier of the query used to register the query handler.
 */
export abstract class BaseQuery implements QueryContract {
  public static readonly queryName: string;

  public get queryName(): string {
    return (this.constructor as typeof BaseQuery).queryName;
  }
}
