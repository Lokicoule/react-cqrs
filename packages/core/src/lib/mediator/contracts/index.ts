import { CommandContract, CommandHandlerContract } from '../../command';
import { EventContract, EventHandlerContract } from '../../event';
import { EventResult } from '../../event/types';
import { QueryContract, QueryHandlerContract } from '../../query';
import { Unsubscribe } from '../../shared/types';

export interface MediatorContract {
  execute<TCommand extends CommandContract, TResult>(
    command: TCommand
  ): TResult;
  dispatch<TEvent extends EventContract>(event: TEvent): EventResult;
  query<TQuery extends QueryContract, TResult>(query: TQuery): TResult;
  registerCommand<TCommand extends CommandContract>(
    command: TCommand,
    handler: CommandHandlerContract<TCommand>
  ): Unsubscribe;
  registerEvent<TEvent extends EventContract>(
    event: TEvent,
    handler: EventHandlerContract<TEvent>
  ): Unsubscribe;
  registerQuery<TQuery extends QueryContract>(
    query: TQuery,
    handler: QueryHandlerContract<TQuery>
  ): Unsubscribe;
}
