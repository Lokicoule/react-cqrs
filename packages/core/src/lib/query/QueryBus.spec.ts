import { QueryBus } from './QueryBus';
import { QueryHandlerContract } from './contracts';
import { QueryAlreadyRegisteredException } from './exceptions/QueryAlreadyRegisteredException';
import { QueryNotFoundException } from './exceptions';
import { BaseQuery } from './models/BaseQuery';

const simpleQuery = {
  queryName: 'query:simple',
};

class SimpleQuery extends BaseQuery {
  public static override readonly queryName = 'query:simple';
}

class PayloadQuery extends BaseQuery {
  public static override readonly queryName = 'query:payload';

  constructor(public readonly payload: string) {
    super();
  }
}

class AsyncHandlerForSimpleQuery
  implements QueryHandlerContract<SimpleQuery, Promise<void>>
{
  public async execute(): Promise<void> {
    return Promise.resolve();
  }
}

class AsyncHandlerForPayloadQuery
  implements QueryHandlerContract<PayloadQuery, Promise<string>>
{
  public execute(query: PayloadQuery): Promise<string> {
    return Promise.resolve(query.payload);
  }
}

class SyncHandlerForSimpleQuery
  implements QueryHandlerContract<SimpleQuery, void>
{
  public execute(): void {
    return;
  }
}

class SyncHandlerForPayloadQuery
  implements QueryHandlerContract<PayloadQuery, string>
{
  public execute(query: PayloadQuery): string {
    return query.payload;
  }
}

function payloadQueryHandler(query: PayloadQuery): string {
  return query.payload;
}

describe('QueryBus', () => {
  let queryBus: QueryBus;

  beforeEach(() => {
    queryBus = new QueryBus();
  });

  describe('register', () => {
    it('should register an async handler for a query', () => {
      const unsubscribe = queryBus.register(
        new SimpleQuery(),
        new AsyncHandlerForSimpleQuery()
      );
      expect(unsubscribe).toBeInstanceOf(Function);
    });

    it('should register a sync handler for a query', () => {
      const unsubscribe = queryBus.register(
        new SimpleQuery(),
        new SyncHandlerForSimpleQuery()
      );
      expect(unsubscribe).toBeInstanceOf(Function);
    });

    it('should register a function as a handler for a query', () => {
      const unsubscribe = queryBus.register(
        new PayloadQuery('payload'),
        payloadQueryHandler
      );
      expect(unsubscribe).toBeInstanceOf(Function);
    });

    it('should register an object as a query', () => {
      const unsubscribe = queryBus.register(
        simpleQuery,
        new AsyncHandlerForSimpleQuery()
      );
      expect(unsubscribe).toBeInstanceOf(Function);
    });

    it('should throw QueryAlreadyRegisteredException when a handler is already registered', () => {
      queryBus.register(new SimpleQuery(), new AsyncHandlerForSimpleQuery());
      expect(() =>
        queryBus.register(new SimpleQuery(), new AsyncHandlerForSimpleQuery())
      ).toThrow(QueryAlreadyRegisteredException);
    });
  });

  describe('execute', () => {
    it('should execute an async handler for a query', async () => {
      queryBus.register(new SimpleQuery(), new AsyncHandlerForSimpleQuery());
      await expect(
        queryBus.execute(new SimpleQuery())
      ).resolves.toBeUndefined();
    });

    it('should execute a sync handler for a query', () => {
      queryBus.register(new SimpleQuery(), new SyncHandlerForSimpleQuery());
      expect(queryBus.execute(new SimpleQuery())).toBeUndefined();
    });

    it('should execute an async handler for a query with payload', async () => {
      queryBus.register(
        new PayloadQuery('payload'),
        new AsyncHandlerForPayloadQuery()
      );
      await expect(queryBus.execute(new PayloadQuery('payload'))).resolves.toBe(
        'payload'
      );
    });

    it('should execute a sync handler for a query with payload', () => {
      queryBus.register(
        new PayloadQuery('payload'),
        new SyncHandlerForPayloadQuery()
      );
      expect(queryBus.execute(new PayloadQuery('payload'))).toBe('payload');
    });

    it('should throw QueryNotFoundException when a handler is not registered', () => {
      expect(() =>
        queryBus.execute(new SimpleQuery())
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('unregister', () => {
    it('should unregister a handler', () => {
      const unsubscribe = queryBus.register(
        new SimpleQuery(),
        new AsyncHandlerForSimpleQuery()
      );
      unsubscribe();
      expect(() => queryBus.execute(new SimpleQuery())).toThrow(
        QueryNotFoundException
      );
    });

    it('should not throw when a handler is not registered', () => {
      const unsubscribe = queryBus.register(
        new SimpleQuery(),
        new AsyncHandlerForSimpleQuery()
      );
      unsubscribe();
      expect(() => unsubscribe()).not.toThrow();
    });
  });
});
