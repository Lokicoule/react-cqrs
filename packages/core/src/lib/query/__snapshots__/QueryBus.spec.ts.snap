// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBus execute should throw QueryNotFoundException when a handler is not registered 1`] = `"Query handler for query:simple not found"`;
